moveit_ros/moveit_servo/src/servo_calcs.cpp

```diff
@@ -736,21 +710,30 @@ void ServoCalcs::enforceSRDFAccelVelLimits(Eigen::ArrayXd& delta_theta)
       // Apply velocity bounds
       if (clip_velocity)
       {
-        // delta_theta = joint_velocity * delta_t
-        const double relative_change = (velocity_limit * parameters_.publish_period) / delta_theta(joint_delta_index);
-        // Avoid nan
-        if (fabs(relative_change) < 1)
-        {
-          delta_theta(joint_delta_index) = relative_change * delta_theta(joint_delta_index);
-          velocity(joint_delta_index) = relative_change * velocity(joint_delta_index);
-        }
+        const double scaling_factor =
+            fabs(velocity_limit * parameters_.publish_period) / fabs(delta_theta(joint_delta_index));
+
+        // Store the scaling factor if it's the smallest yet
+        if (scaling_factor < velocity_limit_scaling_factor)
+          velocity_limit_scaling_factor = scaling_factor;
```

Comment:
Maybe I misunderstood but it looks to me like the function first looks for the smallest velocity limit scaling factor (which can end up being very small) and then, couple lines down, applies that lowest velocity limit to all the joints. What if there is a use case where someone wants, let's say limit the maximum velocity of the largest joint while keeping the other joints moving much faster. Wouldn't this approach then slow the whole robot down since one of the joints is limited to very low velocity?

Reply:
Yeah, the whole robot will slow down.

But, we want the robot to keep moving in the right direction. IMO that trumps speed. Here's a video of what I mean.

![nonlinear_limit_clipping](https://user-images.githubusercontent.com/11284393/91726694-467cd880-eb66-11ea-868f-e1a70886bbd8.gif)